traits/trait_list_object.py

```diff
@@ -53,460 +88,755 @@ def __init__(self, index=0, removed=None, added=None):
-                raise excp
+        if self._should_notify(removed, added):
+            self.notify(norm_index, removed, added)
 
-        self.len_error(len(self) + len(xlist))
+        return removed
 
     def remove(self, value):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            list.remove(self, value)
-            return
-        if trait.minlen < len(self):
-            try:
-                index = self.index(value)
-                removed = [self[index]]
-            except Exception:
-                pass
-
-            list.remove(self, value)
-
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed)
-                )
-        elif len(self) == 0:
-            # Let whatever system error (ValueError) should be raised be
-            # raised.
-            list.remove(self, value)
-        else:
-            self.len_error(len(self) - 1)
+        """ Remove first occurrence of value.
+
+        Parameters
+        ----------
+        value : Any
+            A value contained in the list.
+
+        Raises
+        ------
+        ValueError
+            If the value is not present.
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index between 0 and len(self).
+            added : list
+                Will be []
+            removed : item
+                The removed item.
+
+        """
+        index = self.index(value)
+        self.validate(index, value, [])
+        removed = value
+
+        super().remove(value)
+
+        if self._should_notify(removed, []):
+            self.notify(index, removed, [])
 
     def sort(self, key=None, reverse=False):
-        removed = self[:]
-        list.sort(self, key=key, reverse=reverse)
-        if (
-            getattr(self, "name_items", None) is not None
-            and getattr(self, "trait", None) is not None
-        ):
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed, self[:])
-            )
+        """ Sort the items in the list in ascending order, *IN PLACE*.
+        A custom key function can be supplied to customize
+        the sort order, and the reverse flag can be set to request the result
+        in descending order.
+
+        Parameters
+        ----------
+        key : Callable
+            Custom function.
+        reverse : bool
+            If true, the resulting list will be sorted in descending order.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                The slice between 0 and len(self)
+            added : list
+                Will be [].
+            removed : list
+                Will be []
+
+        """
+
+        self[:] = sorted(self, key=key, reverse=reverse)
+
+        index = slice(0, len(self), None)
+        added = []
+        removed = []
```

Comment:
I think after the discussion that concluded with this: https://github.com/enthought/traits/pull/912#issuecomment-599570303,  only membership changes are meant to trigger validation, Shouldn't it apply to notification too ? Since the other notifications have no notion of element ordering, wouldn't it be odd to make ordering relevant only for `sort` and `reverse`?

Reply:
The comment in https://github.com/enthought/traits/pull/912#issuecomment-599570303 refers to the validation only.
My above comment (14 hours ago) was referring to this: https://github.com/enthought/traits/pull/912#discussion_r390913079:
> The original TraitListObject behaves differently: removed is the original content, and added is the new content. Would this be a backward incompatible change?

The original trait list fires notifications for reordering, with the original ordering list as `removed` and the newly ordered list as `added`.  I don't think we have agreed to change this behaviour, have we?

Reply:
My guess is that a notification is needed in places such as for reordering items shown on a GUI (traitsui).

Reply:
Ok, Made this change.